File: backend/Domain/Lead/ValueObjects/CustomerName.cs
using Domain.Lead.Constants;

namespace Domain.Lead.ValueObjects;

public sealed class CustomerName : IEquatable<CustomerName>
{
    public string Value { get; }

    private CustomerName(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException("Customer name cannot be empty.", nameof(value));

        if (value.Length < LeadConstants.MIN_NAME_LENGTH)
            throw new ArgumentException($"Customer name must be at least {LeadConstants.MIN_NAME_LENGTH} characters.", nameof(value));

        if (value.Length > LeadConstants.MAX_NAME_LENGTH)
            throw new ArgumentException($"Customer name cannot exceed {LeadConstants.MAX_NAME_LENGTH} characters.", nameof(value));

        Value = value.Trim();
    }

    public static CustomerName Create(string value) => new(value);

    public override bool Equals(object? obj) => obj is CustomerName other && Equals(other);

    public bool Equals(CustomerName? other) =>
        other is not null && Value.Equals(other.Value, StringComparison.OrdinalIgnoreCase);

    public override int GetHashCode() => Value.ToLowerInvariant().GetHashCode();

    public override string ToString() => Value;

    public static implicit operator string(CustomerName name) => name.Value;

    public static bool operator ==(CustomerName? left, CustomerName? right) => Equals(left, right);

    public static bool operator !=(CustomerName? left, CustomerName? right) => !Equals(left, right);
}

File: app/lib/features/lead/domain/value_objects/customer_name.dart
import 'package:flutter/foundation.dart';
import '../constants/lead_constants.dart';

@immutable
class CustomerName {
  final String value;

  const CustomerName._(this.value);

  factory CustomerName(String input) {
    final trimmed = input.trim();
    if (trimmed.isEmpty) {
      throw ArgumentError('Customer name cannot be empty');
    }
    if (trimmed.length > LeadConstants.maxNameLength) {
      throw ArgumentError(
        'Customer name cannot exceed ${LeadConstants.maxNameLength} characters',
      );
    }
    return CustomerName._(trimmed);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CustomerName &&
          runtimeType == other.runtimeType &&
          value == other.value;

  @override
  int get hashCode => value.hashCode;

  @override
  String toString() => value;
}